chore: improve debugging for no subnets error in task run

ota42y opened the pull request:
Improve error message to let users know that a tag is required for subnet to run task.

Please read this issue 
https://github.com/aws/copilot-cli/issues/2022

ota42y committed:
fix: add subnets error messages

efekarakus commented on internal/pkg/task/env_runner.go:
```diff
@@ -12,7 +12,7 @@ import (
 
 const (
-	fmtErrPublicSubnetsFromEnv  = "get public subnet IDs from environment %s: %w"
+	fmtErrPublicSubnetsFromEnv  = "get public subnet IDs from environment %s: %w "
```
Can we revert this change? looks like an extra space was added at the end 🙏

ota42y replied:
This is miss, sorry.

efekarakus commented on internal/pkg/task/env_runner.go:
```diff
@@ -60,7 +62,7 @@ func (r *EnvRunner) Run() ([]*Task, error) {
 	}
 	if len(subnets) == 0 {
-		return nil, errNoSubnetFound
+		return nil, fmt.Errorf(fmtErrNoSubnetFoundInEnv, r.Env)
```
Instead of having the error message contain a recommendation "(need %s and %s tags?)" can we use the `log` pkg to add additional context? So the error can be specific but logging can suggest additional actions.

```suggestion
		log.Warningf("Subnets need to be tagged with %s and %s. Are your subnets tagged?\n", deploy.AppTagKey, deploy.EnvTag.Key)
		return nil, errNoSubnetFound
```

Ideally, we would catch this error in the `cli` pkg and do the logging there since this package is meant to be a library pkg.

efekarakus reviewed (commented):
Awesome thank you so much for making debugging this issue better 🙇 ! Some small suggestions below

ota42y committed:
catch error in cli package and show warning

ota42y commented:
I have completed the fix!
I changed to show a warning by error type in cli package so I made error struct public for error check.

efekarakus commented on internal/pkg/cli/task_run.go:
```diff
@@ -438,6 +438,10 @@ func (o *runTaskOpts) runTask() ([]*task.Task, error) {
 	o.spinner.Start(fmt.Sprintf("Waiting for %s to be running for %s.", english.Plural(o.count, "task", ""), o.groupName))
 	tasks, err := o.runner.Run()
 	if err != nil {
+		if err == task.ErrNoSubnetFound {
```
nit: so that wrapped errors are also matched :)
```suggestion
		if errors.Is(err, task.ErrNoSubnetFound) {
```

efekarakus commented on internal/pkg/task/errors.go:
```diff
@@ -9,7 +9,7 @@ import (
 )
 
 var (
-	errNoSubnetFound = errors.New("no subnets found")
+	ErrNoSubnetFound = errors.New("no subnets found")
```
Can we add a top-level comment here since we now have exported errors
```suggestion
// Errors returned while trying to run a task.
var (
	ErrNoSubnetFound = errors.New("no subnets found")
```

efekarakus reviewed (approved):
Yay this looks great! Thank you so much, I left some small nits. Once addressed, we'll remove the "do-not-merge" label which will make [@mergifyio](https://mergify.io/) automatically merge your PR 🚀

ota42y committed:
Update internal/pkg/cli/task_run.go

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

ota42y committed:
add top level comment

ota42y commented:
I fixed it.

However, if you merge https://github.com/aws/copilot-cli/pull/2024 to next version, you must not merge this PR.
Because I changed to use Stack instead of tags in #2024 , so this warning message is no longer needed.
If you don't merge #2024 to next version, this message is useful for users.

efekarakus commented:
Yay sounds good! We'll merge #2024 instead!

ota42y commented:
Ok, I'm going to close this PR when #2024 merged

ota42y commented:
#2024 is merged 🎉
